src/tss2-tcti/tcti-null.c

```diff
@@ -0,0 +1,173 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2025 Juergen Repp
+ * All rights reserved.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h" // IWYU pragma: keep
+#endif
+
+#include <inttypes.h>           // for PRIxPTR, uintptr_t, uint8_t, int32_t
+#include <string.h>             // for NULL, size_t, memset
+
+#include "tcti-common.h"        // for TSS2_TCTI_COMMON_CONTEXT, TCTI_STATE_...
+#include "tss2_common.h"        // for TSS2_RC_SUCCESS, TSS2_RC, TSS2_TCTI_R...
+#include "tss2_tcti.h"          // for TSS2_TCTI_CONTEXT, TSS2_TCTI_INFO
+#include "tss2_tpm2_types.h"    // for TPM2_RC_SUCCESS
+#include "util/aux_util.h"      // for UNUSED
+
+#define LOGMODULE tcti
+#include "tcti-null.h"
+#include "tss2_tctildr.h"       // for Tss2_TctiLdr_Finalize, Tss2_TctiLdr_I...
+#include "util/log.h"           // for LOG_WARNING, LOG_ERROR, LOGBLOB_DEBUG
+
+/*
+ * This function wraps the "up-cast" of the opaque TCTI context type to the
+ * type for the null TCTI context. The only safeguard we have to ensure this
+ * operation is possible is the magic number in the null TCTI context.
+ * If passed a NULL context, or the magic number check fails, this function
+ * will return NULL.
+ */
+TSS2_TCTI_NULL_CONTEXT*
+tcti_null_context_cast (TSS2_TCTI_CONTEXT *tcti_ctx)
+{
+    if (tcti_ctx != NULL && TSS2_TCTI_MAGIC (tcti_ctx) == TCTI_NULL_MAGIC) {
+        return (TSS2_TCTI_NULL_CONTEXT*)tcti_ctx;
+    }
+    return NULL;
+}
+
+/*
+ * This function down-casts the null TCTI context to the common context
+ * defined in the tcti-common module.
+ */
+TSS2_TCTI_COMMON_CONTEXT*
+tcti_null_down_cast (TSS2_TCTI_NULL_CONTEXT *tcti_null)
+{
+    if (tcti_null == NULL) {
+        return NULL;
+    }
+    return &tcti_null->common;
+}
+
+TSS2_RC
+tcti_null_transmit (
+    TSS2_TCTI_CONTEXT *tcti_ctx,
+    size_t size,
+    const uint8_t *cmd_buf)
+{
+    UNUSED(tcti_ctx);
```

Comment:
I would add a LOG_WARNING to each of these _null_ calls so that the user knows something is weird here.